lib/engine.rb

```diff
@@ -8,19 +8,63 @@
 end
 
 module Engine
+  @games = {}
+
+  # Game or Meta
   GAMES = Game.constants.map do |c|
     klass = Game.const_get(c)
-    next if !klass.is_a?(Class) || klass == Game::Base
 
-    klass
+    if c.start_with?('G')
+      if klass.is_a?(Class)
+        klass
+      elsif klass.is_a?(Module)
+        klass::Meta
+      end
+    end
   end.compact
 
-  # Games that are alpha or above
+  # Game or Meta; all that are alpha or above
   VISIBLE_GAMES = GAMES.select { |game| %i[alpha beta production].include?(game::DEV_STAGE) }
 
+  # Game or Meta
   GAMES_BY_TITLE = GAMES.map { |game| [game.title, game] }.to_h
 
+  def self.get_game(game_meta)
+    require_tree "engine/game/#{game_meta.fs_name}" if RUBY_ENGINE == 'opal'
+
+    Engine::Game.constants
+      .map { |c| Engine::Game.const_get(c) }
+      .select { |c| c.constants.include?(:Game) }
+      .map { |c| c.const_get(:Game) }
+      .find { |c| c.title == game_meta.title }
+  end
+
+  # Game
+  def self.game_by_title(title)
+    return @games[title] if @games[title]
+    return @games[title] = GAMES_BY_TITLE[title] if GAMES_BY_TITLE[title].is_a?(Class)
+
+    game_meta = GAMES_BY_TITLE[title]
+    game = get_game(game_meta)
+
+    if game.nil? && RUBY_ENGINE == 'opal'
+      game_by_title(game_meta::DEPENDS_ON) if game_meta::DEPENDS_ON
+
+      src = "/assets/#{game_meta.fs_name}.js"
+
+      # add <script> tag to DOM to load the target game file
```

Comment:
This approach for injecting a `<script>` tag to the DOM to load in the game file as needed doesn't seem to work consistently, though at times I've certainly thought it was consistently working.

I believe one reason why it isn't working regularly is that loading in the remote JS is asynchronous, and the only way to be sure that code is executed once the file has been loaded is with the `onload` callback function, but we can't make `game_by_title()` return what we want form an asynchronous callback.

It probably needs to be removed, and when entering a game by clicking "Enter" on the game card, the new page needs to be loaded in a different way than it is now, so that the server can return prerendered HTML including the `<script>` tag for the given game.

Any help on how to approach this particular problem would be greatly appreciated 😄

Reply:
note: solving this is the main thing needed to solve the unchecked items in the first "TODO" list in the description above

Reply:
Instead of `enter_game`, this should perhaps act more like a normal link would?

https://github.com/tobymao/18xx/blob/e6c6b0bf444f1f3ae033af072f954fe2bd191050/assets/app/view/game_card.rb#L69